src/units/unit.hpp

```diff
@@ -1904,20 +1907,25 @@ class unit : public std::enable_shared_from_this<unit>
 		std::shared_ptr<const unit> parent;
 	};
 
-	recursion_guard update_variables_recursion() const;
-
-
-private:
+	recursion_guard update_variables_recursion(const config& ability) const;
 
 	const std::set<std::string> checking_tags_{"disable", "attacks", "damage", "chance_to_hit", "berserk", "swarm", "drains", "heal_on_hit", "plague", "slow", "petrifies", "firststrike", "poison", "damage_type"};
 	/**
-	 * Check if an ability is active.
+	 * Check if an ability is active. Use recursion checking for prevent crash in UI functions.
 	 * @param ability The type (tag name) of the ability
 	 * @param cfg an ability WML structure
 	 * @param loc The location on which to resolve the ability
 	 * @returns true if it is active
 	 */
 	bool ability_active(const std::string& ability, const config& cfg, const map_location& loc) const;
+	/**
+	 * Check if an ability is active but without use directly recursion_guard checking. This function is called in get_self_ability_bool()and get_adj_ability_bool() for prevent overuse of recursion checking.
```

Comment:
Check if an ability is active. The caller is responsible for preventing excessive recursion, so must hold a recursion_guard.

Reply:
replaced.